citar.el

```diff
@@ -1167,6 +1167,46 @@ With prefix, rebuild the cache before offering candidates."
        (error "Make sure 'citar-library-paths' is a list of paths"))
      (citar--library-file-action key-entry 'attach)))
 
+(defun citar--add-file-to-library (key)
+  "Add a file to the library for KEY.
+The FILE can be added from an open buffer, a file path, or a
+URL."
+  (let* ((source
+          (char-to-string
+           (read-char-choice
+            "Add file from [b]uffer, [f]ile, or [u]rl? " '(?b ?f ?u))))
+         (directory (if (cdr citar-library-paths)
+                        (completing-read "Directory: " citar-library-paths)
+                      (car citar-library-paths)))
+         (file-path
+          ;; Create the path without extension here.
+          (expand-file-name key directory)))
+    (pcase source
+      ("b"
+       (with-current-buffer (read-buffer-to-switch "Add file buffer: ")
+         (let ((extension (file-name-extension (buffer-file-name))))
+           (write-file (concat file-path "." extension) t))))
+      ("f"
+       (let* ((file (read-file-name "Add file: " nil nil t))
+              (extension (file-name-extension file)))
+        (copy-file file
+         (concat file-path "." extension) 1)))
+      ("u"
+       (let* ((url (read-string "Add file URL: "))
+              (extension (url-file-extension url)))
+         (when (< 1 extension)
+           ;; TODO what if there is no extension?
```

Comment:
This is one question that probably needs to be resolved now. Right now, it should just not do anything if given a URL like "https://test.org/one".